### ISSUE TITLE
[proposal] Parameter dim for F.linear (and maybe nn.Linear)

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#52143]
So that we can do linear transforms without doing transposes/permutes/movedims back and forth

**Comment by albanD** [comment]
Just to be sure, the dim argument would tell you which dimension in the input should be used?
And the full function would be equivalent to this?
```
inp = inp.transpose(dim, -1)
out = linear(inp)
out = out.transpose(dim, -1)
```

**Comment by vadimkantorov** [comment]
yep!

**Comment by vadimkantorov** [comment]
E.g. currently we have to use F.conv1d/F.conv2d/F.conv3d and permute the weight (because conv functions can't do unsqueeze to satisfy 1x1 convolution) for simply doing linear transformation

**Comment by vadimkantorov** [comment]
For me this is more of UX question (as dim= lets to hide these transposes under the hood and is supported by a lot of ops)

and then the question of which layouts can be supported by matmul natively without reallocations... (e.g. BCHW layout)

**Comment by vadimkantorov** [comment]
You are right, but the same argument can be made of the dim= feature in general about any op...

about the layouts: maybe this just needs more documentation

E.g. is BCHW physical layout supported by blas? and whether it preserves the layout in the output?(It should be supported by the Conv op, but also by blas? My proposal was uniting both kernels behind the same frontend op)

**Comment by ngimel** [comment]
No, this is not the argument, for reductions with the different `dim` argument  the implementation is materially different, and pytorch provides fast implementations to handle different dims, here in the end the op will still be dispatched to addmm. 
Whatever is supported by addmm/`+` would be supported in nn.Linear
